ES|QL Helper

picandocodigo opened the pull request:
This PR introduces a new ES|QL Helper to the client. The first helper function is to obtain an "Object API" as a result from a query, instead of the default JSON response.

By default, the `query` API returns a Hash response with a `columns` key whose value is an Array of `{ name: type }` Hashes for each column and a `values` key whose value is an Array of Arrays with the values for each row.

This helper function returns an Array of hashes with the columns as keys and the respective values: `{ column['name'] => value }`.

Default response:
```ruby
response = client.esql.query(body: { query: query})

{"columns"=>[
  {"name"=>"@timestamp", "type"=>"date"}, 
  {"name"=>"client.ip", "type"=>"ip"}, 
  {"name"=>"event.duration", "type"=>"long"}, 
  {"name"=>"message", "type"=>"keyword"}, 
  {"name"=>"duration_ms", "type"=>"double"}
], 
"values"=>[
  ["2023-10-23T12:15:03.360Z", "172.21.2.162", 3450233, "Connected to 10.1.0.3", 3.5], 
  ["2023-10-23T12:27:28.948Z", "172.21.2.113", 2764889, "Connected to 10.1.0.2", 2.8], 
  ["2023-10-23T13:33:34.937Z", "172.21.0.5", 1232382, "Disconnected", 1.2], 
  ["2023-10-23T13:51:54.732Z", "172.21.3.15", 725448, "Connection error", 0.7], 
  ["2023-10-23T13:52:55.015Z", "172.21.3.15", 8268153, "Connection error", 8.3], 
  ["2023-10-23T13:53:55.832Z", "172.21.3.15", 5033755, "Connection error", 5.0], 
  ["2023-10-23T13:55:01.543Z", "172.21.3.15", 1756467, "Connected to 10.1.0.1", 1.8]]}
```

Helper response:
```ruby
require 'elasticsearch/helpers/esql_helper'
response = Elasticsearch::Helpers::ESQLHelper.query(client, query)

{"duration_ms"=>3.5, "message"=>"Connected to 10.1.0.3", "event.duration"=>3450233, "client.ip"=>"172.21.2.162", "@timestamp"=>"2023-10-23T12:15:03.360Z"}
{"duration_ms"=>2.8, "message"=>"Connected to 10.1.0.2", "event.duration"=>2764889, "client.ip"=>"172.21.2.113", "@timestamp"=>"2023-10-23T12:27:28.948Z"}
{"duration_ms"=>1.2, "message"=>"Disconnected", "event.duration"=>1232382, "client.ip"=>"172.21.0.5", "@timestamp"=>"2023-10-23T13:33:34.937Z"}
{"duration_ms"=>0.7, "message"=>"Connection error", "event.duration"=>725448, "client.ip"=>"172.21.3.15", "@timestamp"=>"2023-10-23T13:51:54.732Z"}
{"duration_ms"=>8.3, "message"=>"Connection error", "event.duration"=>8268153, "client.ip"=>"172.21.3.15", "@timestamp"=>"2023-10-23T13:52:55.015Z"}
{"duration_ms"=>5.0, "message"=>"Connection error", "event.duration"=>5033755, "client.ip"=>"172.21.3.15", "@timestamp"=>"2023-10-23T13:53:55.832Z"}
{"duration_ms"=>1.8, "message"=>"Connected to 10.1.0.1", "event.duration"=>1756467, "client.ip"=>"172.21.3.15", "@timestamp"=>"2023-10-23T13:55:01.543Z"}

picandocodigo commented:
run docs-build

szabosteve commented on docs/helpers.asciidoc:
```diff
@@ -160,3 +160,79 @@ end
 ----
 --
+
+[discrete]
```
I'd add an anchor to this section so it'll be easier to link to it. I can imagine it'll be a popular read. :) 
```suggestion
[discrete]
[[esql-helper]]
```

szabosteve commented on docs/helpers.asciidoc:
```diff
@@ -160,3 +160,79 @@ end
+[discrete]
+=== ES|QL Helper
+
+This helpers provides an Object response from the ESQL `query` API instead of the default JSON value.
```
```suggestion
This helper provides an object response from the ESQL `query` API instead of the default JSON value.
```

szabosteve commented on docs/helpers.asciidoc:
```diff
@@ -160,3 +160,79 @@ end
+]}
+----
+
+The helper returns an Array of hashes with the columns as keys and the respective values. So for the previous example, it would return the following:
```
```suggestion
The helper returns an array of hashes with the columns as keys and the respective values. So for the previous example, it would return the following:
```

szabosteve commented on docs/helpers.asciidoc:
```diff
@@ -160,3 +160,79 @@ end
+]
+----
+
+Additionally, a block can be specified to work on the response data. Pass in a block to `query` and it will yield each item in the Array of responses.
```
I'm not sure 'transform' is the right word, but if I understand correctly, the block helps you display the data differently.
```suggestion
Additionally, a block can be specified to transform the response data. Pass in a block to `query` and it will yield each item in the array of responses.
```

szabosteve reviewed (approved):
Docs look good! Left a couple of suggestions, mostly nits. Please take or leave them!

ioanatia commented on elasticsearch/spec/integration/helpers/esql_helper_spec.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  it 'can work on the iterated objects when a block is specified' do
+    response = esql_helper.query(client, query).each do |r|
```
I think we are actually passing a block to `Array#each` not ` Elasticsearch::Helpers::ESQLHelper.query`.

picandocodigo commented:
I changed the implementation to @ioanatia's idea to pass in Procs as "parsers" to the helper. I've also updated the documentation and added examples to the source code docs, so that the API reference is updated and includes a couple of code examples.

picandocodigo commented:
The buildkite failure is unrelated to this change:
* https://github.com/elastic/elasticsearch/pull/106663

picandocodigo commented:
I'm going to merge this and warn that it's experimental, so I can backport it for the 8.13 release of the client. We can iterate over it or do breaking changes while it's marked as experimental.
